url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/dnssec/algorithms/JavaSecSignatureVerifier.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2015-2016 the original author or authors
+ *
+ * This software is licensed under the Apache License, Version 2.0,
+ * the GNU Lesser General Public License version 2 or later ("LGPL")
+ * and the WTFPL.
+ * You may choose either license to govern your use of this software only
+ * upon the condition that you accept all of the terms of either
+ * the Apache License 2.0, the LGPL 2.1+ or the WTFPL.
+ */
+package de.measite.minidns.dnssec.algorithms;
+
+import de.measite.minidns.dnssec.DNSSECValidationFailedException;
+import de.measite.minidns.dnssec.SignatureVerifier;
+
+import java.security.InvalidKeyException;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.Signature;
+import java.security.SignatureException;
+
+public abstract class JavaSecSignatureVerifier implements SignatureVerifier {
+    private final KeyFactory keyFactory;
+    private final String signatureAlgorithm;
+
+    public JavaSecSignatureVerifier(String keyAlgorithm, String signatureAlgorithm) throws NoSuchAlgorithmException {
+        keyFactory = KeyFactory.getInstance(keyAlgorithm);
+        this.signatureAlgorithm = signatureAlgorithm;
+
+        // Verify signature algorithm to be valid
+        Signature.getInstance(signatureAlgorithm);
+    }
+
+    public KeyFactory getKeyFactory() {
+        return keyFactory;
+    }
+
+    @Override
+    public boolean verify(byte[] content, byte[] rrsigData, byte[] key) {
+        try {
+            PublicKey publicKey = getPublicKey(key);
+            Signature signature = Signature.getInstance(signatureAlgorithm);
+            signature.initVerify(publicKey);
+            signature.update(content);
+            return signature.verify(getSignature(rrsigData));
+        } catch (NoSuchAlgorithmException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)